Rewrite acount meta processing loop

otselnik opened the pull request:
Simplified account processing loop in account storage creating function

otselnik committed:
account_storage: for loop -> while loop in "new"

otselnik committed:
fixes

otselnik committed:
Merge remote-tracking branch 'origin/develop' into process_accounts_meta_slice

s-medvedev commented on evm_loader/program/src/account_storage.rs:
```diff
@@ -29,75 +29,65 @@ impl<'a> ProgramAccountStorage<'a> {
-                    AccountData::Account(ref acc) => acc,
-                    _ => return Err(ProgramError::InvalidAccountData),
-                };
+                let account = AccountData::get_account(&account_data)?;
```
Why do you use static function calls instead of member-function?

s-medvedev commented on evm_loader/program/src/account_storage.rs:
```diff
@@ -29,75 +29,65 @@ impl<'a> ProgramAccountStorage<'a> {
-                accounts.push(Some(sol_account));
-            } else {
-                accounts.push(None);
+                aliases.push((sol_account.get_ether(), index));
```
You can use `accounts.len()` instead of `index` here.

s-medvedev commented on evm_loader/program/src/account_storage.rs:
```diff
@@ -29,75 +29,65 @@ impl<'a> ProgramAccountStorage<'a> {
-                    debug_print!("code_account == Pubkey::new_from_array([0u8; 32])");
+                    debug_print!("Common account");
+
+                    if caller_id == H160::zero() {
+                        caller_id = account.ether;
+                        debug_print!("caller id: {}", &caller_id.to_string());
+                    }
```
`caller_id` will be the first non-code account. What happens if the first account doesn't belong to program_id? Or with code?

I suggest we need strictly limit the order of caller and contract accounts (as described in `Call` instruction).
Moreover, the contract always should contain the code part and the caller should be **without** code.

otselnik committed:
remove useless parameter from apply function

otselnik committed:
use AccountStorage trait to acquire ethereum addresses of contact and caller

otselnik committed:
move caller account checks from entrypoint

otselnik committed:
ProgramAccountStorage::new
make own function to create solidity  account containing contract
move contact and caller accounts acquiring from loop
use loop to get other ethereum accounts

otselnik committed:
add docs

otselnik committed:
correct error

otselnik committed:
fix checks

otselnik committed:
clarify description

otselnik committed:
add missed check

otselnik committed:
move clock account acquiring to account_storage

otselnik committed:
Merge remote-tracking branch 'origin/develop' into process_accounts_meta_slice

otselnik committed:
fix merge
